Fix subsasgn behaviour when logical indices are used.

genuinelucifer opened the pull request:
Fixes issue #598
When logical indices are used, pass the valid indices (if any). If there are
no valid indices then return the input as is.

genuinelucifer commented:
Hopefully there aren't any hidden cases which would break by this. As far as I have tested this seems a reasonable solution.

genuinelucifer commented:
There's a test failure in subsasgn.. When I ran the test in the file, I just got a warning of failure which I stupidly ignored. I'll fix it tomorrow.

latot commented:
Hi, can you add this tests plis? i know this seems a lot, but every test confirm differents parts of this, although likes similars, and this need have this minimum size of the matrix to can test all of this:
```
%!shared a, b
%! a = [1 2 3 5;4 5 6 9;7 5 3 2];
%! b = sym(a);

%!function test_bool (a, b, c)
%!  assert (a(c), double (b(c)))
%!  a(c) = 0;
%!  b(c) = 0;
%!  assert (nnz (a == 0), nnz (b == 0))
%!endfunction

%!test
%! c = false;
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)

%!test
%! c = [false true];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false; true];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false true false true; true false true false; false true false true];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false true false true false];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false; true; false; true; false];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false true; false true];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false true false; false true false];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)

%!test
%! c = [false true; false true; true false];
%! test_bool (a, b, c)
%! test_bool (a, b, c | true)
%! test_bool (a, b, c & false)
```
I think the issue will be solved when all this test pass.

Thx. Cya.

latot commented:
Only as a note, this need a lot of test because the behavior is very different with logic index and non-logic index..........

genuinelucifer commented:
@latot I added your tests with following 2 changes:  
* I removed the first line `assert (a(c), double (b(c)))` from `test_bool` as it exposed the problem with `subsref` and not `subsasgn`. I will fix the problems with `subsref` in next PR.
* I removed 2 of the tests where you mixed logicals with integers in `c` as this is not supported in octave for any type of matrix anyway

latot commented:
Hi!, i'm agree to remove the subsref tests, we will need open a new issue for subsref, its possible keep the assert as a comment? only to remember it.

The test mixing logicals and integers is supported for octave:
```
octave:15> a=[1, 2, 3;4, 5, 6]
a =

   1   2   3
   4   5   6

a([true 1;true 2])=0
a =

   0   2   3
   0   5   6
```
But i think we can do it later.

Thx. Cya.

genuinelucifer commented:
I also removed some of your tests which were superseded by some other tests.  
  
> The test mixing logicals and integers is supported for octave

I see. It works only if all the logicals are `true`. We can probably support this then. I'll check.

latot commented:
mm, some test are lost...., was this intentional?

latot commented:
comment removed, let me check something...

genuinelucifer commented on inst/@sym/private/mat_replace.m:
```diff
@@ -28,6 +28,17 @@
+    end
+    if (~isscalar(b) && ~isvector(b))
+      warning('OctSymPy:subsagn:rhs_shape', ...
+            'B neither vector nor scalar in indexed A(I)=B: unusual, did you intend this?')
```
@cbm755 Is this warning important?

cbm755 replied:
I'm not sure.  If Octave double allows it, I guess we should too.

genuinelucifer replied:
Octave double allows it without any warning. Should I remove the warning then? (from here and from the original file `mat_mask_asgn`)

genuinelucifer commented:
5 doctests are failing because `evalpy` is deprecated!!.. I'll see if I need to rebase!

latot commented:
I'm with the same problem!, should that be fixed from master and merge here?

cbm755 commented:
Ok, I reverted the commit for evalpy, sorry about that.

cbm755 commented:
It might not have been that commit, it might've been the new doctest 0.5.0 coming in, which needed changes.  Anyway, master is fine again on CI and I've restarted jobs.

genuinelucifer commented:
@latot I have checked with different values and saw that octave actually replaces all the `true` in indices to `1` if the indices consist of integers too! And `false` is not at all allowed in this case. Thanks for the test cases...  
  
I also saw that currently it was an error if the index used had more than 2 rows because of how `ind2subs` works. So, I fixed that in here too.
  
@cbm755 I think this is now ready. Please review it. Also, I think the 3 commits are fine. Should I squash the first 2 or is it ok like this?

cbm755 commented:
Also: please add yourself to the copyright header of these two files

latot commented on inst/@sym/subsasgn.m:
```diff
@@ -513,8 +513,78 @@
+%! test_bool (a, b, c & false)
+
+%!test
+%! c = [2 true; true 3; true true];
```
Hi, octave support doubles and bolean mixed for index, i think is a valid test.

cbm755 replied:
No it doesn't.  `[true 2]` is cast to double.  `whos` will show you.

latot replied:
uu, true, when we mix logical with doubles octave transform logical values to that type.

mmm, i think we need to be careful and check how this works when we send a symbolic index matrix.

cbm755 replied:
Good point, file another issue so we don't loose it?

genuinelucifer replied:
After this PR is merged I am planning to fix issues related to subsref too.  Actually quite similar issues are there with subsref too!

genuinelucifer commented on inst/@sym/subsasgn.m:
```diff
@@ -513,8 +513,78 @@
 %! assert( isequal( a, b));
 
+%!test
+%! % indexing with a dimension more than 2
```
I actually got confused while choosing the label. This test actually is for indexing with a mXn matrix where either one of m or n is more than 2. So probably something along the lines of "indexing with a matrix with size of a dimension more than 2"!?

cbm755 replied:
Ok, or "2D indexing with length in one dimension more than 2".

genuinelucifer commented on inst/@sym/subsasgn.m:
```diff
@@ -513,8 +513,78 @@
+%! a = sym ([1, 2; 3, 4; 5 6]);
+%! indices = [1 2; 3 4; 5 6];
+%! b = [10 11; 12 13; 14 15];
+%! a (indices) = b;
```
Oh thanks! Will do.

genuinelucifer commented on inst/@sym/subsasgn.m:
```diff
@@ -513,8 +513,78 @@
+%!function test_bool (a, b, c)
+%!  a(c) = 0;
+%!  b(c) = 0;
+%!  assert (nnz (a == 0), nnz (b == 0))
```
That would work too. I thin this function,  in current state, just checks for the changes in a, b and not the exact values.
Since @latot wrote it, probably he can comment here!

latot replied:
Hi, note, i only choose nnz because there was 2 options, transform all the matrix to double, or count the number of zeros, i choose the last, if you think can help the other option no problem changing it.

Note 2, how the input matrix don't have any zero, count the number of zeros is the same as check the values changed in the matrices.

Thx. Cya.

cbm755 replied:
Hmm `nnz` is much faster than `double` currently.  Speeding up double would be a good idea, easier after pytave I'm sure, so I wasn't going to bother.

Actually just calling `isequal(A,B)` is twice as fast as converting to double (this will convert them both to sym).  Maybe that's the way to go here.

genuinelucifer commented:
@cbm755 the extra warnings were due to a mistake on my side. I issued the warning even when rhs was empty. I've fixed it now.
.
I think we should drop the warning as octave's double inherently allows subsasgn with rhs as a matrix.
.
Also, I have fixed the spacing issues now. Please review it.

genuinelucifer commented on inst/@sym/subsasgn.m:
```diff
@@ -517,4 +518,71 @@
 %! a(1, 2) = [];
 
+
+%% Tests from mat_replace
```
I have added this as al the tests here on after mostly relate to the changes done in mat_replace only.

genuinelucifer commented on inst/@sym/private/mat_replace.m:
```diff
@@ -66,12 +79,7 @@
-    z = mat_mask_asgn(A, subs{1}, b);
-    return
-
-  elseif (length(subs) == 1 && strcmp(subs{1}, ':') && length(b) == 1)
```
I have remove the first if as that condition is now taken care at the start of this method.

genuinelucifer commented on inst/@sym/private/mat_replace.m:
```diff
@@ -28,6 +29,18 @@
+      z = A;
+      return;
+    end
+    if (~ isscalar (b) && ~ isvector (b) && ~ isempty (b))
```
Added an isempty condition to block the warnings in that case.

cbm755 commented:
LGTM!  Merging.

Can we drop `private/mat_mask_asgn` now?
